From 214a1e4ea099f3dd99011d142cba0ca5715ef175 Mon Sep 17 00:00:00 2001
From: Daniel Bolgheroni <Daniel.Bolgheroni@windriver.com>
Date: Mon, 16 Dec 2024 16:31:34 -0300
Subject: [PATCH] Add support for multiple hosts in a daemonset

A daemonset overrides template generates a separate
daemonset and configuration secret for each host, even when
configurations are identical. This approach leads to unnecessary
resource consumption and complexity.

This adds support for specifying multiple hosts for a single daemonset
in conf.overrides.neutron_dhcp-agent.hosts[].name.

Change-Id: I80e87174addd26af12490c4f18cae8316c2bd670
Signed-off-by: Daniel Bolgheroni <Daniel.Bolgheroni@windriver.com>
---
 .../templates/utils/_daemonset_overrides.tpl  | 50 ++++++++++++++++---
 1 file changed, 44 insertions(+), 6 deletions(-)

diff --git a/helm-toolkit/templates/utils/_daemonset_overrides.tpl b/helm-toolkit/templates/utils/_daemonset_overrides.tpl
index 40359f0f..7bf0dd23 100644
--- a/helm-toolkit/templates/utils/_daemonset_overrides.tpl
+++ b/helm-toolkit/templates/utils/_daemonset_overrides.tpl
@@ -35,9 +35,23 @@ limitations under the License.
               {{- $current_dict := dict }}
 
               {{/* set daemonset name */}}
-              {{/* Note: long hostnames can cause the 63 char name limit to be
-              exceeded. Truncate the hostname if hostname > 20 char */}}
-              {{- if gt (len $host_data.name) 20 }}
+              {{/* There are cases where the name can be a list instead of a single
+                   value. Since some metadata are defined using these names as the base
+                   name, create a hostgroup. */}}
+              {{- $hostgroup_name := "" }}
+              {{- $host_data_names := list }}
+              {{- if eq (kindOf $host_data.name) "slice" }}
+                {{- range $host_data.name }}
+                  {{- $host_data_names = append $host_data_names . }}
+                {{- end }}
+                {{- $host_data_names_all := join "-" $host_data_names }}
+                {{- $hostgroup_name = printf "hostgroup-%s" (sha256sum $host_data_names_all | trunc 7) }}
+              {{- end }}
+              {{- if $hostgroup_name }}
+                {{- $_ := set $current_dict "name" $hostgroup_name }}
+              {{- /* Note: long hostnames can cause the 63 char name limit to be
+                     exceeded. Truncate the hostname if hostname > 20 char */}}
+              {{- else if gt (len $host_data.name) 20 }}
                 {{- $_ := set $current_dict "name" (substr 0 20 $host_data.name) }}
               {{- else }}
                 {{- $_ := set $current_dict "name" $host_data.name }}
@@ -60,7 +74,14 @@ limitations under the License.
               {{- $_ := set $nodeSelector_dict "key" "kubernetes.io/hostname" }}
               {{- $_ := set $nodeSelector_dict "operator" "In" }}
 
-              {{- $values_list := list $host_data.name }}
+              {{- $values_list := list }}
+              {{- if eq (kindOf $host_data.name) "slice" }}
+                {{- range $host_data.name }}
+                  {{- $values_list = append $values_list . }}
+                {{- end }}
+              {{- else }}
+                {{- $values_list = list $host_data.name }}
+              {{- end }}
               {{- $_ := set $nodeSelector_dict "values" $values_list }}
 
               {{- $list_aggregate := list $nodeSelector_dict }}
@@ -123,7 +144,16 @@ limitations under the License.
                     {{- $_ := set $label_dict "key" "kubernetes.io/hostname" }}
                     {{- $_ := set $label_dict "operator" "NotIn" }}
 
-                    {{- $values_list := list $host_data.name }}
+                    {{/* - $values_list := list $host_data.name */}}
+                    {{- $values_list := list }}
+                    {{- if eq (kindOf $host_data.name) "slice" }}
+                      {{- range $host_data.name }}
+                        {{ $values_list = append $values_list . }}
+                      {{- end }}
+                    {{- else }}
+                      {{- $values_list = list $host_data.name }}
+                    {{- end }}
+
                     {{- $_ := set $label_dict "values" $values_list }}
 
                     {{- $list_aggregate := append $context.Values.__current_label.matchExpressions $label_dict }}
@@ -164,7 +194,15 @@ limitations under the License.
               {{- $_ := set $default_dict "key" "kubernetes.io/hostname" }}
               {{- $_ := set $default_dict "operator" "NotIn" }}
 
-              {{- $values_list := list $host_data.name }}
+              {{- $values_list := list }}
+              {{- if eq (kindOf $host_data.name) "slice" }}
+                {{- range $host_data.name }}
+                  {{ $values_list = append $values_list . }}
+                {{- end }}
+              {{- else }}
+                {{- $values_list = list $host_data.name }}
+              {{- end }}
+
               {{- $_ := set $default_dict "values" $values_list }}
 
               {{- $list_aggregate := append $context.Values.__default.matchExpressions $default_dict }}
-- 
2.34.1

